### PR TITLE
auto connect a ipywidget slider to a linear selector

### PR DESCRIPTION
**Pull request by kushalkolar**
implements #297 for int and float sliders

@minel-arinel 

Usage:
```python
# make ipywidget
ipywidget_slider = IntSlider(min=0, max=100, description="slider")

# connect to selector
selector.add_ipywidget_handler(ipywidget_slider)
```

### CODE REVIEW

**Comment by kushalkolar** [comment]
ipywidget -> selector is working well, selector -> ipywidget is not working well if the step size is not 1, it's wonky and jumpy

**Comment by kushalkolar** [comment]
ok this works now, made an attr `_moving` for `BaseSelector` :D

**Comment by kushalkolar** [comment]
Implemented for `LinearRegionSelector` as well, and made `limits` a settable proprety for both `LinearRegionSelector` and `LinearSelector`. 

@clewis7 will merge after your review

https://github.com/fastplotlib/fastplotlib/assets/9403332/449bbebc-097f-4a42-86cf-51fc98c59947

https://github.com/fastplotlib/fastplotlib/assets/9403332/b2a841cd-e51d-4325-a01f-05db6d2113d4

File: fastplotlib/graphics/selectors/_linear.py [hunk 1]
```diff
@@ -18,6 +19,19 @@
 
 
 class LinearSelector(Graphic, BaseSelector):
+    @property
+    def limits(self) -> Tuple[float, float]:
+        return self._limits
+
+    @limits.setter
+    def limits(self, values: Tuple[float, float]):
+        if len(values) != 2 or not all(map(lambda v: isinstance(v, Real), values)):
```
Comment (clewis7) [inline fastplotlib/graphics/selectors/_linear.py:28]: add code comment here for `not all(map(lambda v: isinstance(v, Real), values))`

pretty ambiguous as is, comment will be helpful later to remember what is happening

File: fastplotlib/graphics/selectors/_linear.py [hunk 2]
```diff
@@ -18,6 +19,19 @@
...
+            raise TypeError(
+                "limits must be an iterable of two numeric values"
+            )
+        self._limits = tuple(map(round, values))  # if values are close to zero things get weird so round them
```
Comment (clewis7) [inline fastplotlib/graphics/selectors/_linear.py:32]: where does `round` come from? I don't see any definitions or imports

File: fastplotlib/graphics/selectors/_linear.py [hunk 3]
```diff
@@ -84,7 +94,8 @@ def __init__(
         if len(limits) != 2:
             raise ValueError("limits must be a tuple of 2 integers, i.e. (int, int)")
 
-        limits = tuple(map(round, limits))
+        self._limits = tuple(map(round, limits))
```
Comment (clewis7) [inline fastplotlib/graphics/selectors/_linear.py:97]: see above

File: fastplotlib/graphics/selectors/_linear_region.py [hunk 4]
```diff
@@ -9,6 +18,19 @@
 
 
 class LinearRegionSelector(Graphic, BaseSelector):
+    @property
+    def limits(self) -> Tuple[float, float]:
+        return self._limits
+
+    @limits.setter
+    def limits(self, values: Tuple[float, float]):
+        if len(values) != 2 or not all(map(lambda v: isinstance(v, Real), values)):
```
Comment (clewis7) [inline fastplotlib/graphics/selectors/_linear_region.py:27]: same as above

**Review by clewis7 — COMMENTED** [review]
mostly looks good to me, I think just some more code comments would be helpful for understanding what is going on

From what I see, and correct me if I am wrong...

A user will create an ipywidget slider and then can add this to a linear selector using the "add" method; internally you set up the widget to have bidirectional linkage with the associated linear selector

A given selector looks like it can have more than one associated ipywidget, which are managed by a private list

File: fastplotlib/graphics/selectors/_linear.py (hunk 1 above)
Comment (kushalkolar) [reply fastplotlib/graphics/selectors/_linear.py:28]: checks that `values` is an iterable of two real numbers

**Comment by kushalkolar** [comment]
> A user will create an ipywidget slider and then can add this to a linear selector using the "add" method; internally you set up the widget to have bidirectional linkage with the associated linear selector

Yes, they add a handler. 

> A given selector looks like it can have more than one associated ipywidget, which are managed by a private list

yes

**Pull request merged**